### PR TITLE
extend coverage of package specs

### PR DESCRIPTION
**Pull request by vpereira**

### CODE REVIEW

File: src/api/spec/features/webui/packages_spec.rb [hunk 1]
```diff
@@ -115,6 +115,25 @@
...
+    let(:target_package) { package.name }
+    let!(:bs_request) { create(:bs_request, source_project: source_project,
+                                           target_project: target_project,
+                                           target_package: target_package) }
+    scenario 'see a request' do
+      login user
+      visit package_show_path(package: package, project: user.home_project)
+      click_link('Requests')
+      expect(page).to have_css('table#all_requests_table tr', count:1)
+      # binding.pry
```
Comment (Ana06) [inline src/api/spec/features/webui/packages_spec.rb]: :eyes:

**Comment by vpereira** [comment]
I would love to have a tag to "DO NOT REVIEW YET". To have a PR open is the best way to have two or more colleagues contributing to the same branch, have it being tested by the CI and etc,  as I'm doing with @bgeuken.

**Comment by coolo** [comment]
well, in other projects we just prefix the PR title with WIP

**Comment by hennevogel** [comment]
> I would love to have a tag to "DO NOT REVIEW YET"

The point is explaining yourself. So when you add the existing "DO NOT MERGE" label, explain why you do in a comment. We require that BTW.

**Comment by vpereira** [comment]
> The point is explaining yourself. So when you add the existing "DO NOT MERGE" label, explain why 
> you do in a comment. We require that BTW.

However the "DO NOT MERGE" is just to avoid to merge "not ready" PR, but it can still be reviewed. To save some reviewer time, the WIP as suggested by @coolo  should do the trick.

**Comment by hennevogel** [comment]
> save some reviewer time

Again, this is just about communication. If you add that no one needs to review this in the description/comment when you add the label everything is fine. We already have too many labels.

**Comment by Ana06** [comment]
@vpereira 

> I would love to have a tag to "DO NOT REVIEW YET". To have a PR open is the best way to have two or more colleagues contributing to the same branch, have it being tested by the CI and etc, as I'm doing with @bgeuken.

You can work together in a branch without open a PR. You could for example to add it to the Trello card to make it more visible. As @hennevogel, I don't think we need another label, but you should explain the reason why you add _DO NOT MERGE_. :wink: 

@coolo 

> well, in other projects we just prefix the PR title with WIP

I would do that to specify that it is not finish, but to receive some feedback. :thinking:

**Comment by Ana06** [comment]
So, can we already review it or why is this open?

**Comment by vpereira** [comment]
@bgeuken, @coolo  do you think that this method https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/webui/package_controller.rb#L495 should be a before_filter?

**Comment by vpereira** [comment]
https://trello.com/c/CbtMOS3j/1236-p8-package-creation

**Comment by vpereira** [comment]
please I would like to have it merge before that we start the card https://trello.com/c/BdbyI2EW/834-p11-bootstrap-packagerequests, the other missing tests can be pushed either in another PR or by the one doing the migration from package#requests to bootstrap. @openSUSE/open-build-service

**Comment by DavidKang** [comment]
@vpereira, why is needed this PR been merged before [Package#request](https://trello.com/c/BdbyI2EW/834-p11-bootstrap-packagerequests)? Is there any dependency?

File: src/api/spec/features/webui/packages_spec.rb [hunk 2]
```diff
@@ -115,6 +117,30 @@
     expect(find('#flash-messages')).to have_text('Package was successfully removed.')
   end
 
+  describe 'existing requests' do
+    let(:source_project) { create(:project_with_package, name: 'source_project') }
+    let(:source_package) { source_project.packages.first }
+    let(:target_project) { user.home_project }
+    # let(:target_package) { package.name }
```
Comment (DavidKang) [inline src/api/spec/features/webui/packages_spec.rb]: :eyes:

File: src/api/spec/features/webui/packages_spec.rb [hunk 3]
```diff
@@ -1,6 +1,8 @@
 require 'browser_helper'
 require 'webmock/rspec'
 
+# CONFIG['global_write_through'] = true
```
Comment (mdeniz) [inline src/api/spec/features/webui/packages_spec.rb]: We don't that anymore, please remove it :+1:

File: src/api/app/controllers/webui/package_controller.rb [hunk 4]
```diff
@@ -103,6 +103,7 @@ def dependency
       redirect_back(fallback_location: project_show_path(project: params[:dproject]))
       return
     end
+
```
Comment (mdeniz) [inline src/api/app/controllers/webui/package_controller.rb]: I guess this commit's message is totally unrelated to what changes you added...?? :thinking:

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 5]
```diff
@@ -1501,4 +1502,64 @@ def do_request(params)
...
+    end
+
+    context 'valid package name' do
+      before do
+        post :save_new, params: { project: source_project, name: 'new-package',
+                                  title: 'package foo',
+                                  description: 'awesome package foo' }
+      end
+
+      it { expect(response).to have_http_status(:redirect) }
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: Use better the `redirect_to` method and add the path to where is being redirected.

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 6]
```diff
@@ -1501,4 +1502,64 @@ def do_request(params)
...
+      it { expect(flash[:notice]).to be_present }
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: This is unneeded as you ask in the test below for it having a value already

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 7]
```diff
@@ -1501,4 +1502,64 @@ def do_request(params)
...
+
+    context 'invalid package name' do
+      before do
+        post :save_new, params: { project: source_project, name: 'A' * 250,
+                                  title: 'package foo',
+                                  description: 'awesome package foo' }
+      end
+
+      it { expect(response).to have_http_status(:redirect) }
+      it { expect(flash[:error]).to be_present }
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: Same here

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 8]
```diff
@@ -1501,4 +1502,64 @@ def do_request(params)
...
+    end
+
+    context 'invalid package name' do
+      before do
+        post :save_new, params: { project: source_project, name: 'A' * 250,
+                                  title: 'package foo',
+                                  description: 'awesome package foo' }
+      end
+
+      it { expect(response).to have_http_status(:redirect) }
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: same here

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 9]
```diff
@@ -1501,4 +1502,64 @@ def do_request(params)
...
+      it { expect(flash[:error]).to be_present }
+      it { expect(flash[:error]).to match("Invalid package name:\s.*") }
+    end
+  end
+
+  describe 'GET #dependency' do
+    let(:my_arch) { Architecture.find_by(name: 'i586') }
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: [Nitpick] Why not using `i586` as the name of the architecture?

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 10]
```diff
@@ -1501,4 +1502,64 @@ def do_request(params)
...
+      allow(Backend::Api::BuildResults::Binaries).to receive(:fileinfo_ext).and_return(nil)
+      get :dependency, params: { project: source_project,
+                                 arch: my_arch.name, package: source_package,
+                                 dproject: target_project.name,
+                                 drepository: repo_for_target_project.name,
+                                 repository: repo_for_target_project.name,
+                                 filename: 'foo.txt' }
+    end
+
+    it { expect(response).to have_http_status(:redirect) }
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: same here

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 11]
```diff
@@ -1501,4 +1502,64 @@ def do_request(params)
...
+    it { expect(flash[:error]).to be_nil }
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: Better asking for not present?

File: src/api/spec/cassettes/Packages/existing_requests/see_a_request.yml [hunk 12]
```diff
@@ -0,0 +1,1118 @@
+---
```
Comment (mdeniz) [inline src/api/spec/cassettes/Packages/existing_requests/see_a_request.yml]: This commit's message is wrong, those files are just cassettes recorded... not factories.

File: src/api/spec/features/webui/packages_spec.rb (hunk 3 above)
Comment (vpereira) [reply src/api/spec/features/webui/packages_spec.rb]: why we don't need it anymore?

Comment (mdeniz) [reply src/api/spec/features/webui/packages_spec.rb]: We do it now in the `spec/support/vcr.rb` file instead of spread in the spec files.

**Comment by vpereira** [comment]
@mdeniz do you want to review it?

File: src/api/spec/features/webui/packages_spec.rb (hunk 2 above)
Comment (vpereira) [reply src/api/spec/features/webui/packages_spec.rb]: could you please re-review it @DavidKang

**Comment by bgeuken** [comment]
> @bgeuken, @coolo do you think that this method https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/webui/package_controller.rb#L495 should be a before_filter?


Sorry, I overlooked this one before.

The file line doesn't seem to match anymore. I suspect you meant `check_package_name_for_new`. Yes, this looks like a before filter to me. (Though the first condition could also be a model validation and the third one an authorization. But one thing at a time^^)

**Comment by vpereira** [comment]
@bgeuken the right line is https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/webui/package_controller.rb#L487, i just wrote him, because he was the author and you because we were working together on the card.. but i just moved, together with @mdeniz  it to a before filter :)

**Comment by bgeuken** [comment]
Ah!:smiley: I missed the change because the link was pointing to the master branch

File: src/api/spec/features/webui/packages_spec.rb (hunk 2 above)
Comment (DavidKang) [reply src/api/spec/features/webui/packages_spec.rb]: Please, remove this comment and also [vcr comment](https://github.com/openSUSE/open-build-service/pull/5527/files#diff-fafd54efb3601a2010161498e1fefb6fR4)

**Comment by vpereira** [comment]
@hennevogel  please merge it as soon its green, since the card, is set to done https://trello.com/c/CbtMOS3j/710-p8-package-creation

**Comment by ChrisBr** [comment]
> please merge it as soon its green, since the card, is set to done 

This is not how it works! Cards are done when the definition of done is met. One item on it the PR is merged. So I would say the card is not done ...

**Comment by vpereira** [comment]
@ChrisBr  hey Chris, DUNNO. The card *is* set to done, the PR was linked to that, so I'm assuming that everything was done, reviewed and etc.. no?

**Comment by ChrisBr** [comment]
>The card is set to done, the PR was linked to that, so I'm assuming that everything was done, reviewed and etc.. no?

If this PR belongs to this card, then the card is not done, no? I moved the card back into review now!

**Comment by vpereira** [comment]
> The card is set to done, the PR was linked to that, so I'm assuming that everything was done, reviewed and etc.. no?
> 
> If this PR belongs to this card, then the card is not done, no? I moved the card back into review now!

Ok thanks. I don't know, I was in FTO until yesterday, not sure what happened here in this meantime ;)

File: src/api/spec/features/webui/packages_spec.rb (hunk 3 above)
Comment (mdeniz) [inline src/api/spec/features/webui/packages_spec.rb]: Please remove this comment... don't forget that :smiling_imp: .

File: src/api/spec/features/webui/packages_spec.rb (hunk 2 above)
Comment (mdeniz) [inline src/api/spec/features/webui/packages_spec.rb]: Remove this also :smiling_imp:

**Comment by vpereira** [comment]
> Please squash the commit 7a25fcc with e1c3c38

done

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 13]
```diff
@@ -1635,4 +1631,74 @@ def do_request(params)
...
+    let(:package_name) { 'new-package' }
+    let(:my_user) { user }
+    let(:post_params) do
+      { project: source_project, name: package_name,
+                              title: 'package foo',
+                              description: 'awesome package foo' }
+    end
+
+    context 'Package#save failed' do
+      let(:package_name) { 'new-package' }
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: I think that this is not needed, as you already have it in the line 1636 already

**Comment by mdeniz** [comment]
Great Job :clap:

**Pull request merged**